### DISCUSSION TITLE
dhcp plugin sending DHCP requests on wrong interface

### DISCUSSION THREAD

**Discussion by lowjoel** [strongswan/strongswan#1558] [Q&A] [answered]
Hello there, I've got Strongswan and dnsmasq running on the same router on OpenWrt with multiple VLANs. Each VLAN has a separate instance of dnsmasq running (because of a different DNS suffix) and they are all running using `--bind-interfaces` which uses `SO_BINDTODEVICE`. I'm trying to recreate the "roadwarrior" configuration for my setup, starting with a Windows client over EAP-RADIUS/Network Policy Server (which uses PEAP under the hood). I mostly followed the [Windows Client](https://docs.strongswan.org/docs/5.9/interop/windowsEapServerConf.html) guide which got me a working Phase1/Phase2 set up, except that I can't get routes and IP addresses working using the dhcp plugin.

When configuring my connection using IP pools that Strongswan manages, I can send DNS queries to the router no problem. But connections beyond the router doesn't seem to be working, probably indicating yet more of a routing problem, but that's tangential to the problem at hand (that I may get to later.)

The logs show that the DHCP requests are being sent/arriving on the wrong interface on dnsmasq:

```
Sat Feb 25 09:59:10 2023 daemon.info ipsec: 06[CFG] sending DHCP DISCOVER for <client mac> to 192.168.1.1
Sat Feb 25 09:59:10 2023 daemon.warn dnsmasq-dhcp[1]: no address range available for DHCP request via <wan vlan, router wan IP>
Sat Feb 25 09:59:11 2023 daemon.info ipsec: 06[CFG] sending DHCP DISCOVER for <client mac> to 192.168.1.1
Sat Feb 25 09:59:11 2023 daemon.warn dnsmasq-dhcp[1]: no address range available for DHCP request via <wan vlan, router wan IP>
Sat Feb 25 09:59:11 2023 daemon.info ipsec: 09[MGR] ignoring request with ID 12, already processing
Sat Feb 25 09:59:12 2023 daemon.info ipsec: 10[MGR] ignoring request with ID 12, already processing
Sat Feb 25 09:59:13 2023 daemon.info ipsec: 06[CFG] sending DHCP DISCOVER for <client mac> to 192.168.1.1
Sat Feb 25 09:59:13 2023 daemon.warn dnsmasq-dhcp[1]: no address range available for DHCP request via <wan vlan, router wan IP>
Sat Feb 25 09:59:15 2023 daemon.info ipsec: 07[MGR] ignoring request with ID 12, already processing
Sat Feb 25 09:59:16 2023 daemon.info ipsec: 06[CFG] sending DHCP DISCOVER for <client mac> to 192.168.1.1
Sat Feb 25 09:59:16 2023 daemon.warn dnsmasq-dhcp[1]: no address range available for DHCP request via <wan vlan, router wan IP>
Sat Feb 25 09:59:20 2023 daemon.info ipsec: 06[CFG] sending DHCP DISCOVER for <client mac> to 192.168.1.1
Sat Feb 25 09:59:20 2023 daemon.warn dnsmasq-dhcp[1]: no address range available for DHCP request via <wan vlan, router wan IP>
Sat Feb 25 09:59:22 2023 daemon.info ipsec: 15[MGR] ignoring request with ID 12, already processing
```

I've already set the interface and server address (both the lan vlan's broadcast, as well as the router's own unicast IP):

```
dhcp {

    # Always use the configured server address.
    force_server_address = yes

    # Derive user-defined MAC address from hash of IKE identity and send client
    # identity DHCP option.
    # identity_lease = no

    # Interface name the plugin uses for address allocation.
    interface = br-lan.1024

    # Whether to load the plugin. Can also be an integer to increase the
    # priority of this plugin.
    load = yes

    # DHCP server unicast or broadcast IP address.
    server = <both 192.168.1.1 and 192.168.1.255 tried>

    # Use the DHCP server port (67) as source port when a unicast server address
    # is configured.
    # use_server_port = no

}
```

I've been digging through the source code, and I'm wondering if this is because `SO_BINDTODEVICE` only affects receiving packets? man(7) socket:

```
SO_BINDTODEVICE
              Bind this socket to a particular device like “eth0”, as
              specified in the passed interface name.  If the name is an
              empty string or the option length is zero, the socket
              device binding is removed.  The passed option is a
              variable-length null-terminated interface name string with
              the maximum size of IFNAMSIZ.  If a socket is bound to an
              interface, only packets received from that particular
              interface are processed by the socket.  Note that this
              works only for some socket types, particularly AF_INET
              sockets.  It is not supported for packet sockets (use
              normal [bind(2)](https://www.man7.org/linux/man-pages/man2/bind.2.html) there).

              Before Linux 3.8, this socket option could be set, but
              could not retrieved with [getsockopt(2)](https://www.man7.org/linux/man-pages/man2/getsockopt.2.html).  Since Linux 3.8,
              it is readable.  The optlen argument should contain the
              buffer size available to receive the device name and is
              recommended to be IFNAMSIZ bytes.  The real device name
              length is reported back in the optlen argument.
```

This is OpenWrt 22.03.3 (strongswan 5.9.5).

Maybe there's some other option that I'm missing. Any help would be appreciated.

**Comment by lowjoel** [comment]
Oh, it seems that the DHCP plugin is setting the giaddress of the DHCP Discover message to the WAN IP of the router 🤔. If I set the server address to 255.255.255.255 (the only broadcast address that the plugin recognises at the moment), dnsmasq returns with a DHCPOFFER but charon says that `DHCP DISCOVER timed out`.

It's not the cond timeout bug fixed in the newer versions of strongswan -- I see that dnsmasq returns DHCPOFFER 3 times before the timeout message.

**Comment by lowjoel** [comment, marked as answer]
OK I've spent more time tracing the netlink code and it seems that it uses the `interfaces_use` list for Charon also affects the DHCP plugin. In that scenario, since I've only allowed the WAN interfaces to accept IKE traffic, Strongswan uses the default route's source IP as the correct giaddress.

In that sense, the initial question has been answered.

However, I'm not sure if  this is expected - it does seem to be surprising behaviour. Should this be documented or should the behaviour be changed?

**Reply by tobiasbrunner** [reply]
Interesting. Yes, the plugin uses `kernel_interface_t::get_source_addr` to determine the source address/route to the DHCP server, which, as you noticed, is affected by the interface config (although not always, see below). It would probably make sense to use a different mechanism here (maybe connecting a new socket, which is something e.g. `ping` does to determine the source IP to a particular destination). Please open an issue for this.

Note that if `kernel-netlink.fwmark` is set (e.g. to `!42`) so a certain mark is excluded from the routing table where strongSwan installs its routes, the _kernel-netlink_ plugin uses a more efficient source address/route lookup (without dumping all matching routes). And I just noticed that this will bypass the interface filter (AFAIK we can't tell the kernel to ignore certain interfaces in a `RTM_GETROUTE` request). If the best route goes via an ignored interface, no source address/route will be found as there are no alternatives returned by the kernel. On the other hand, the kernel's route lookup gives users more control over the routing so maybe ignoring interfaces only on the receive side is OK in that case.

**Reply by lowjoel** [reply]
Thanks Tobias, #1573
